### PR TITLE
registry/memory: add support for domain options

### PR DESCRIPTION
**Pull request by ben-toogood**
Add support for domain options in the memory registry.

### CODE REVIEW

File: registry/memory/memory.go [hunk 1]
```diff
@@ -34,27 +35,32 @@ type Registry struct {
...
+	// records can be passed for testing purposes
 	records := getServiceRecords(options.Context)
 	if records == nil {
 		records = make(map[string]map[string]*record)
 	}
 
 	reg := &Registry{
 		options:  options,
-		records:  records,
+		records:  make(map[string]services),
```
Comment (domwong) [inline registry/memory/memory.go]: Not using `records`? Then you can probably delete the block above

File: registry/memory/memory.go [hunk 2]
```diff
@@ -192,93 +218,199 @@ func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption
...
+	// if the nodes not empty, we replace the version in the store and exist, the rest of the logic
+	// is cleanup
+	if len(version.Nodes) > 0 {
+		m.records[options.Domain][s.Name][s.Version] = version
+		return nil
+	}
+
+	// if this version was the only version of the service, we can remove the whole service from the
+	// registry and exit
+	if len(versions) == 0 {
```
Comment (domwong) [inline registry/memory/memory.go]: When would `len(versions)` be 0? We haven't deleted this version from `versions` so not sure when this is true.

File: registry/memory/memory.go [hunk 3]
```diff
@@ -192,93 +218,199 @@ func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption
...
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = defaultDomain
+	}
+
+	// if it's a wildcard domain, return from all domains
+	if options.Domain == registry.WildcardDomain {
+		m.Lock()
```
Comment (domwong) [inline registry/memory/memory.go]: Can probably be a `RLock` instead

File: registry/memory/memory.go [hunk 4]
```diff
@@ -192,93 +218,199 @@ func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption
...
+	for _, o := range opts {
+		o(&options)
+	}
+	if len(options.Domain) == 0 {
+		options.Domain = defaultDomain
+	}
+
+	// if it's a wildcard domain, list from all domains
+	if options.Domain == registry.WildcardDomain {
+		m.Lock()
```
Comment (domwong) [inline registry/memory/memory.go]: Can probably be an RLock instead

**Comment by ben-toogood** [comment]
Thanks for the review @domwong. I've actioned all your points.

**Review by domwong — APPROVED** [review]

**Pull request merged**